### PR TITLE
nixos/hardware/intel: add opencl option

### PR DESCRIPTION
**Pull request by Mic92**
###### Description of changes

Part of https://github.com/NixOS/nixos-hardware/pull/573

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Kiskae** [comment]
OpenCL on intel is messy, I believe this is the support table as it currently exists:

| pkg | CPU | GPU | Notes |
| ---  | --- | --- | --- |
| `intel-compute-runtime` | no | yes | since `Broadwell` 5th gen, open source |
| `intel-ocl` | yes | no[^1] | since `Ivy Bridge` 3rd gen, closed source, deprecated |
| `beignet` | no | yes | since `Ivy Bridge` 3rd gen, open source, deprecated |
| `intel-opencl-runtime` | yes[^2] | no | newer releases are only available as an SDK, not available in `nixpkgs` |

[^1]: package actually includes GPU driver, but the ICD isn't installed properly so it doesn't work.
[^2]: according to the arch wiki, it supports non-intel CPUs as well.

**Comment by Mic92** [comment]
Thanks for the table! Should we make `intel-ocl` opt-in if unfree packages are allowed and have an option that we can turn off in future?

**Comment by Kiskae** [comment]
That is probably the best approach, perhaps another option to use `beignet` on older GPUs?

**Comment by SuperSandro2000** [comment]
> That is probably the best approach, perhaps another option to use `beignet` on older GPUs?

That then belongs into nixos-hardware propably.

File: nixos/modules/hardware/video/intel.nix [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{ config, lib, pkgs, ... }:
+{
+  options.hardware.intel.opencl.enable = lib.mkEnableOption (lib.mdDoc
+    "intel opencl runtime (Install icd for Intel GPUs)"
+  );
+
+  config = lib.mkIf config.hardware.intel.opencl.enable {
+    hardware.opengl.extraPackages = with pkgs; [
+      intel-compute-runtime
+      intel-ocl
```
Comment (SuperSandro2000) [inline nixos/modules/hardware/video/intel.nix:10]: Having that under video is not fully correct, as it also has parts that only use the CPU, it is more general intel hardware.

**Review by SuperSandro2000 — COMMENTED** [review]
we should also update https://github.com/NixOS/nixpkgs/blob/6f558d3009c20cd2decc5b36057f78a033a6ff11/nixos/doc/manual/configuration/gpu-accel.chapter.md#intel-sec-gpu-accel-opencl-intel and possible https://github.com/NixOS/nixpkgs/blob/6f558d3009c20cd2decc5b36057f78a033a6ff11/nixos/modules/hardware/opengl.nix#L119

**Review by JohnRTitor — COMMENTED** [review]
@Mic92 could you rebase this and address the issues raised?

File: nixos/modules/hardware/cpu/intel-opencl.nix [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  options.hardware.intel.opencl.runtime = lib.mkOption {
+    type = lib.types.nullOr (
+      lib.types.enum [
+        "intel-compute-runtime"
+        "intel-ocl"
+        "beignet"
+      ]
+    );
+    default = null;
+    description = lib.mdDoc ''
```
Comment (JohnRTitor) [inline nixos/modules/hardware/cpu/intel-opencl.nix:17]: Please remove the occurrence(s) of `lib.mdDoc`

**Review by JohnRTitor — DISMISSED** [review]

File: nixos/modules/hardware/cpu/intel-opencl.nix (hunk 2 above)
Comment (Mic92) [reply nixos/modules/hardware/cpu/intel-opencl.nix]: I was confused by grepping the old checkout here...

File: nixos/modules/hardware/cpu/intel-opencl.nix [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    type = lib.types.nullOr (
+      lib.types.enum [
+        "intel-compute-runtime"
+        "intel-ocl"
+        "beignet"
+      ]
+    );
+    default = null;
+    description = ''
+      Select the Intel OpenCL runtime to use. Choose your runtime based on your Intel CPU/GPU generation.
```
Comment (Mic92) [inline nixos/modules/hardware/cpu/intel-opencl.nix:18]: I consider this option documentation, also it's the same amount of code to add stuff to extraPackages, this one makes it easier to discover what needs to be done to get opencl running on a given Intel CPU.

**Comment by Mic92** [comment]
> we should also update https://github.com/NixOS/nixpkgs/blob/6f558d3009c20cd2decc5b36057f78a033a6ff11/nixos/doc/manual/configuration/gpu-accel.chapter.md#intel-sec-gpu-accel-opencl-intel and possible
> 
> https://github.com/NixOS/nixpkgs/blob/6f558d3009c20cd2decc5b36057f78a033a6ff11/nixos/modules/hardware/opengl.nix#L119

Done. Anything else?

**Review by JohnRTitor — APPROVED** [review]
LGTM, but this probably is too minimal to have its own module

I agree with https://github.com/NixOS/nixos-hardware/issues/992#issuecomment-2180265087 and I feel like small modules like this can be merged into a big and general intel module.

**Review by Lyndeno — APPROVED** [review]

**Comment by Mic92** [comment]
> LGTM, but this probably is too minimal to have its own module
> 
> I agree with [NixOS/nixos-hardware#992 (comment)](https://github.com/NixOS/nixos-hardware/issues/992#issuecomment-2180265087) and I feel like small modules like this can be merged into a big and general intel module.

If we would make a cut and ignore anything older than broadwell, this could probably just be part of an igpu module instead? I am not sure how relevant opencl was in previous generations...

**Comment by Lyndeno** [comment]
Perhaps this module could be moved as-is to gpu? Like [nixos/modules/services/hardware/amdgpu.nix](https://github.com/NixOS/nixpkgs/blob/00d80d13810dbfea8ab4ed1009b09100cca86ba8/nixos/modules/services/hardware/amdgpu.nix)

**Comment by Mic92** [comment]
> Perhaps this module could be moved as-is to gpu? Like [nixos/modules/services/hardware/amdgpu.nix](https://github.com/NixOS/nixpkgs/blob/00d80d13810dbfea8ab4ed1009b09100cca86ba8/nixos/modules/services/hardware/amdgpu.nix)
> 
> Edit: Oops, amdgpu is not under gpu directory, but maybe it should be?

Sounds good. If someone wants to take over, feel free. I might get back to this later otherwise.

**Pull request closed without merging**